### PR TITLE
[go] Bump images, dependencies and versions to go 1.23rc2

### PR DESCRIPTION
**Pull request by cpanato**
#### What type of PR is this?

/kind feature
/hold

#### What this PR does / why we need it:

- Bump images, dependencies and versions to go 1.23rc2

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/release/issues/3650

/assign @MadhavJivrajani @liggitt @puerco @xmudrii  @saschagrunert 
cc @kubernetes/release-managers 

#### Does this PR introduce a user-facing change?

```release-note
Kubernetes is now built with go 1.23rc2
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MadhavJivrajani** [comment]
https://github.com/kubernetes/kubernetes/pull/125926 should help with those failures, I'm going to try and get that green today/tomorrow so that we can rebase this.

**Comment by MadhavJivrajani** [comment]
@cpanato whenever you get the chance, could you please rebase the PR here? https://github.com/kubernetes/kubernetes/pull/125926 is merged so the lint errors should go away now.

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -538,6 +538,7 @@ EOF
   local go_version
   IFS=" " read -ra go_version <<< "$(GOFLAGS='' go version)"
   local minimum_go_version
+  # TODO: Update to go1.23 as soon we are ready to merge this
   minimum_go_version=go1.22
```
Comment (MadhavJivrajani) [inline hack/lib/golang.sh:542]: @cpanato are we good to change this?

Comment (cpanato) [reply hack/lib/golang.sh:542]: i think we change that when the 1.23 is official released not in rcs

**Comment by liggitt** [comment]
This lgtm. I'm not sure when in August go1.23 is expected to cut, but I'm ok with merging this speculatively to flush out any issues we have with it, and dropping back to latest go1.22 if go1.23.0 isn't released by our cut date. Should we wait until after code freeze to merge this?

**Comment by MadhavJivrajani** [comment]
> This lgtm. I'm not sure when in August go1.23 is expected to cut, but I'm ok with merging this speculatively to flush out any issues we have with it, and dropping back to latest go1.22 if go1.23.0 isn't released by our cut date. Should we wait until after code freeze to merge this?

@liggitt thinking more about this - there was a release update email from the Go team today that said go1.23 is ~4 weeks out ([email](https://groups.google.com/d/msgid/golang-dev/99106a8a-7221-4426-ac03-9313e241f5den%40googlegroups.com?utm_medium=email&utm_source=footer)). 4 weeks from today (18th July) is 15th August. But it looks like v1.31.0 cut date is 13th Aug, so it seems unlikely that go1.23 will be out by then unless we want to delay the cut of v1.31.0 for any reason. If that were to happen, we can probably take a call re:bumping to go1.30 later at that point. But we can probably hold off merging this for now, and I can kickstart testing with go1.23rc2 in the meantime. 

cc @dims

**Comment by liggitt** [comment]
oof... so close... definitely bump to rc2 in this PR

I'd even *consider* merging rc2 to HEAD even if we planned to go back to 1.22.x for the release, just to get signal on post-submits for a couple weeks

**Comment by dims** [comment]
> I'd even consider merging rc2 to HEAD even if we planned to go back to 1.22.x for the release

Can we setup a couple of jobs that runs 1.22.x to ensure we don't inadvertently merge 1.23 specific stuff? https://tip.golang.org/doc/go1.23#language

Once we have them, then i'd support moving to rc2 .. two way door essentially!

**Comment by cpanato** [comment]
> > I'd even consider merging rc2 to HEAD even if we planned to go back to 1.22.x for the release
> 
> Can we setup a couple of jobs that runs 1.22.x to ensure we don't inadvertently merge 1.23 specific stuff? https://tip.golang.org/doc/go1.23#language
> 
> Once we have them, then i'd support moving to rc2 .. two way door essentially!

updated to rc2 :)

**Comment by MadhavJivrajani** [comment]
> I think go1.22.x in our go.mod files keeps that from happening

Yep this is true.

From https://go.dev/doc/modules/gomod-ref:
> The go directive affects use of new language features:
>
> For packages within the module, the compiler rejects use of language features introduced after the version specified by the go directive. For example, if a module has the directive go 1.12, its packages may not use numeric literals like 1_000_000, which were introduced in Go 1.13.

/hold cancel

**Comment by liggitt** [comment]
sgtm, add an issue labeled as release-blocking and in the milestone to either promote to 1.23.0 or revert this to 1.22.x before our 1.31 release

**Pull request merged**